Update DistributedBag with improvements for DFS

Guillaume-Helbecque opened the pull request:
This PR follows #21958, and adds a variant of the `DistributedBag` module, designed for unbalanced depth-first tree-search at scale (more details are provided in #21958). 

The main purposes of this PR are the following:
- deprecate the existing `DistributedBag` module, and rename it `DistributedBagDeprecated` (see https://github.com/chapel-lang/chapel/issues/21958#issuecomment-1819666763);
- upload the new module, and make it the new `DistributedBag`;
- add a set of correctness tests: `add`, `addBulk`, `clear`, `writeThis`, and `nqueens`. The latter gets closer to a real-life application, and implements a parallel distributed backtracking algorithm to solve small N-Queens instances. The number of explored nodes as well as the number of solutions found are checked.

Some performance tests are also planned, but will be added in a future PR.

In addition to those tests, this module has already been used to solve real-world large scale optimization problems (including flowshop scheduling and knapsack) and led to scientific publications (conference and journal papers). I can provide references for those interested. Recently, the DS has been used in experiments totalling up to 400 computer nodes (51,200 CPU cores).

Resolves #21958.

Future Work Suggested by @mppf:
 * Look for ways that programmers might not need to pass the task ID. Perhaps #16405 could be of use.

Guillaume-Helbecque commented:
@mppf The PR is ready for a first round. I added some more docs to facilitate as much as possible the review.

mppf commented on modules/packages/DistributedBag.chpl:
```diff
@@ -326,958 +357,809 @@ module DistributedBag {
-      return (true, elem);
-    }
+                    if hasElt {
+                      stolenElts.insert(0, elt);
```
I suspect that such distributed list insertion is not particularly efficient. The `list` data type works better if you are appending; inserting at the `0`th position is `O(n)`. Worth at least a TODO comment.

Guillaume-Helbecque replied:
Added a TODO

mppf reviewed (approved):
Thanks for putting this PR together!

After you address the feedback we will still need to do testing with the PR. Sometimes that reveals an issue. For now, I will wait to begin such testing until the review feedback is addressed.

lydia-duncan commented on modules/packages/DistributedBagDeprecated.chpl:
```diff
@@ -0,0 +1,1286 @@
+  _______
+*/
+
+@deprecated("The DistributedBagDeprecated module is deprecated")
```
I would still consider Michael the primary reviewer for this PR, but I saw the commit that added this file go past and wanted to point out that while deprecating and renaming a module at the same time will mean future uses of it are appropriately warned for, it does not alert current users that things have changed out from under them.

Since this module was not already considered stable, I think it is probably okay, but in the future I would recommend a slightly slower process of change, to ensure that users have warning and can opt-in to the new behavior instead of being surprised when things have changed.

mppf replied:
@lydia-duncan - that's something to take up with myself and @bradcray since @Guillaume-Helbecque here is just doing what I/we advised in https://github.com/chapel-lang/chapel/issues/21958#issuecomment-1819501798

Guillaume-Helbecque commented on modules/packages/DistributedBag.chpl:
```diff
@@ -83,183 +94,191 @@
-    there are larger numbers of elements. The better the locality, the better raw
-    performance and easier it is to redistribute work.
+    The maximum capacity of each segment. This is crucial to ensure memory usage
+    does not grow out of control.
```
If the limit is reached, we cease to offer more, and return the number of elements successfully inserted. It means that some elements may be missed. This behaviour is inherited from the old data structure, itself inherited from the `Collection` module.

This behaviour is described in the doc of the `addBulk` proc:

>  Insert elements in bulk in segment ``taskId``. If the bag instance rejects
      an element (e.g., if :const:distributedBagMaxSegmentCap is reached), we
      cease to offer more. We return the number of elements successfully inserted.

and in that case, some warnings are generated at runtime:

> warning("maximum capacity reached: some elements may have been missed.");

mppf replied:
OK. IMO it is preferable to rely on the return value from `add` and `addBulk` instead of the runtime warning for this. See also my comment https://github.com/chapel-lang/chapel/pull/23204/files#r1412532304 .

Guillaume-Helbecque commented on modules/packages/DistributedBag.chpl:
```diff
@@ -326,958 +357,809 @@ module DistributedBag {
-      updates across nodes, and may miss elements resulting from any concurrent
-      insertion or removal operations.
+      :return: Depending on the scenarios: `(true, elt)` if we successfully removed
+               element `elt` from DistBag; `(false, defaultOf(eltType))` otherwise.
```
That's right! Added a TODO for that.

Guillaume-Helbecque commented on modules/packages/DistributedBag.chpl:
```diff
@@ -326,958 +357,809 @@ module DistributedBag {
+
+              dom = {0..#segment.nElts};
+              for i in dom {
+                buffer[i] = segment.block.elts[segment.block.headId + i];
```
Yes, any global operations on a `DistBag`, such as `getSize`, `these`, etc, may miss elements due to concurrent updates of the data structure. Those methods implement a "snapshot" approach (inherited from the old `DistBag`), leading to non-deterministic outputs in most cases.

I tried to take care of this by adding the following warning in the doc of those procs, but it seems that I forgot this one.

> This method is best-effort and can be non-deterministic for concurrent
        updates across nodes, and may miss elements resulting from any concurrent
        insertion or removal operations.

mppf replied:
Right, but I think in this case it's not using a snapshot at all & the "best effort" words would imply that the program will run but might miss some updates. But I would expect accessing uninitialized memory if there were concurrent removals to the segment.

Guillaume-Helbecque replied:
OK, I got it. I think I mixed up both approaches. To fix this, I disabled concurrent updates while iterating, and restored the old doc for `these`.

Guillaume-Helbecque commented on modules/packages/DistributedBag.chpl:
```diff
@@ -326,958 +357,809 @@ module DistributedBag {
+    /*
+      Iterate over the segments/locales eligible to be stolen from, according to
+      the specified policy. By default, the random strategy is chosen and the
+      calling thask/locale cannot be chosen. We can specify how many tries we
```
You're right. I added more details.

Guillaume-Helbecque commented on modules/packages/DistributedBag.chpl:
```diff
@@ -326,958 +357,809 @@ module DistributedBag {
+      calling thask/locale cannot be chosen. We can specify how many tries we
+      want, by default, only 1 is performed.
+    */
+    iter victim(const N: int, const callerId: int, const policy: string = "rand",
```
Added a TODO.

Guillaume-Helbecque commented on modules/packages/DistributedBag.chpl:
```diff
@@ -18,62 +17,74 @@
+  the work stealing (WS) paradigm. In few words, when a segment is (near)-empty,
+  the associated task will try to steal elements from the public portion of other
+  segments, potentially from other locales. During WS operations, thieve tasks
+  synchronize using a lock. This mechanism is transparently managed by the remove
```
Implemented option 2

Guillaume-Helbecque commented on modules/packages/DistributedBag.chpl:
```diff
@@ -18,62 +17,74 @@
+                     elements locally, i.e. from another segment of its bag instance.
+                     To select the victim task, we iterate over the segments using
+                     either the "random" (default) or "round-robin" policy. When a
+                     segment is eligible to be stolen from, one element is removed.
```
Once we steal one from another segment, the element is returned and the stealing process stops. I clarified this point in this comment.

Guillaume-Helbecque commented on modules/packages/DistributedBag.chpl:
```diff
@@ -83,183 +94,191 @@
   */
   pragma "always RVF"
-  record DistBag : serializable {
+  record DistBag : serializable
```
It could be nice to apply this style to the remaining package modules in a follow-up PR. What do you think?

mppf replied:
Probably best to wait until doing other changes to one of those package modules. But, if you would like to update them, I think we would accept PRs in that area. Please just make separate PRs for your work on each.

Guillaume-Helbecque commented on modules/packages/DistributedBag.chpl:
```diff
@@ -83,183 +94,191 @@
     var bag = new DistBag(int, targetLocales=ourTargetLocales);
 
+  DistBag supports a set of basic methods applying to the segment indexed by
+  ``taskId`` in ``{0..<here.maxTaskPar}``. This guarantees the local DFS ordering.
```
OK to use your sentence, but I replaced "accept" by "require", because "accept" sounds like "you can set `taskId` or not" to me, while it is required.

Guillaume-Helbecque replied:
If a wrong segment ID is given, we can distinguish two cases:
- `segmentID >= here.maskTaskPar`: this produces `error: halt reached - array index out of bounds`, since segments are internally mapped using the domain `{0..<here.maskTaskPar}`. We could consider supporting user-defined domains, but I think the current implementation is not too restrictive.
- `segmentID < here.maskTaskPar` and `segmentID != taskID`: if the "wrong" calls are done concurrently, this could cause race conditions because the "private" region of each segment is not parallel-safe, and methods like `add` or `remove` act on it. Otherwise, I would say it just work with a different local segment without issue.

Guillaume-Helbecque replied:
To be more precise, the only condition to ensure parallel-safety of DistBag (and also DFS) is to map **one segment per task**. It is not mandatory to map segment 0 on task 0, segment 1 on task 1, and so on. Segment 0 on task 1, segment 1 on task 2 and so on is also valid, like any other permutation.

mppf replied:
I like the words in your last comment. Could you make sure they are in a documentation comment somewhere?

Guillaume-Helbecque commented on modules/packages/DistributedBag.chpl:
```diff
@@ -326,958 +357,809 @@ module DistributedBag {
+        when "rand" {
+          var id: int;
+          var victims: [0..#N] int = noinit;
+          permutation(victims);
```
That's a good point! I marked your comments as TODOs.

Guillaume-Helbecque commented on modules/packages/DistributedBag.chpl:
```diff
@@ -18,62 +17,74 @@
-  it avoids the overhead of remotely accessing class fields, bounding performance
-  on communication.
+ /*
+  A highly parallel segmented multi-pool specialized for depth-first search (DFS).
```
Added "Implementation Description:" at the top

Guillaume-Helbecque commented on modules/packages/DistributedBag.chpl:
```diff
@@ -83,183 +94,191 @@
+  While the bag is safe to use in a distributed manner, each locale always operates
+  on its privatized instance. This means that it is easy to add data in bulk,
+  expecting it to be distributed, when in reality it is not; if another locale needs
+  data, it will steal work on-demand.
```
Added "Here is an example of concurrent operations on DistBag across multiple locales and tasks"

Guillaume-Helbecque commented:
Thank you @mppf and @lydia-duncan for your valuable reviews (and sorry for the many typos). I made the requested changes, and marked some TODO that will be addressed in a follow-up PR, if this is still OK for you.

Guillaume-Helbecque commented on modules/packages/DistributedBag.chpl:
```diff
@@ -326,958 +357,809 @@ module DistributedBag {
-            deallocate(buffer);
-          }
+          // Create a snapshot
+          var block = segment.block;
```
Same as above: I disabled concurrent updates while iterating.

mppf commented:
@Guillaume-Helbecque - Just to let you know, we have feature freeze for Chapel 1.33 tomorrow. I'm not expecting this PR to go in by then, but it also means that it might be a week before we are able to merge it.

Thanks for making the improvements here. I'll get back to looking at this PR soon.

mppf commented on test/library/packages/DistributedBag/add.chpl:
```diff
@@ -0,0 +1,20 @@
+use DistributedBag;
+
+config param nElems = 3;
+config param defaultMaxSegmentCap = 1024*1024;
```
Wouldn't we want this to evolve along with the distributedBagMaxSegmentCap in the module?

For example, you could write `const defaultMaxSegmentCap = DistributedBag.distributedBagMaxSegmentCap`.

Guillaume-Helbecque replied:
Exactly

mppf commented on test/library/packages/DistributedBag/add.good:
```diff
@@ -0,0 +1,4 @@
+add.chpl:17: warning: maximum capacity reached: some elements may have been missed.
+add.chpl:17: warning: maximum capacity reached: some elements may have been missed.
+add.chpl:17: warning: maximum capacity reached: some elements may have been missed.
```
what is printing this warning? Generally, it's best to avoid a warning printed from a library, and rely on other mechanisms instead. Is there anything we could do instead of this warning? I thought that `add` would return a smaller number if something is not added? Could we argue that the warning isn't necessary and it's rather the calling code (the test in this case) that should be checking the number returned by `bag.add`?

Guillaume-Helbecque replied:
I strengthened the test by checking that `add` returns `true` when it is expected, and by checking that we get the good number of failures.

mppf commented on test/library/packages/DistributedBag/addBulk.chpl:
```diff
@@ -0,0 +1,22 @@
+use DistributedBag;
+
+config param nElems = 1000;
+config param defaultMaxSegmentCap = 1024*1024;
```
See the other comment about this pattern

mppf commented on test/library/packages/DistributedBag/addBulk.good:
```diff
@@ -0,0 +1,2 @@
+addBulk.chpl:18: warning: maximum capacity reached: some elements may have been missed.
```
See the other comment about this pattern

mppf commented on test/library/packages/DistributedBag/nqueens.chpl:
```diff
@@ -0,0 +1,232 @@
+
+        // Check termination
+        if hasWork {
+          eachTaskState[taskId].write(BUSY);
```
It might improve performance to write `ref curTaskState = eachTaskState[taskId]` just outside of the `while true` so that the array lookup does not need to be re-done each time through here. Same for `eachLocaleState`. Of course, this is not worth bothering with if you don't care about this performance of this test.

Guillaume-Helbecque replied:
This test doesn't deal with performance, but any minor modification that can make the code more optimized is welcomed.

mppf commented on test/library/packages/DistributedBag/nqueens.compopts:
```diff
@@ -0,0 +1 @@
+--fast
```
Why is `--fast` added here? In our nightly testing, we have different configurations that test with and without `--fast`. Is this test too slow to be practical if `--fast` is not given?

mppf replied:
I think that 28s is acceptable & we can add intermediate flags like `-O` later if it becomes too slow. I think it's important that it run with bounds checking on in some configurations, and `--fast` disables bounds checking.

mppf commented on test/unstable/packageModules.chpl:
```diff
@@ -12,6 +12,7 @@ import CopyAggregation;
 import Crypto;
 import Curl;
 import DistributedBag;
+import DistributedBagDeprecated;
```
This should be a test in `test/deprecated/` somewhere rather than in `test/unstable` (because DistributedBagDeprecated is deprecated rather than unstable). I would expect you can simply add a test that includes `import DistributedBagDeprecated;` and put the warning in the `.good` file.

Guillaume-Helbecque replied:
This test checks the `unstable` warnings of the package modules, so I added this line in order to reflect the new `DistributedBagDeprecated` package module. For some reasons, the deprecation warning is shown in `.good` instead of the unstable one, like for the `BinaryIO` module. 

Nevertheless, I agree that a standalone deprecation test is needed.

Guillaume-Helbecque replied:
The deprecation test has already been added in a previous commit.

mppf replied:
IMO it makes sense to remove `DistributedBagDeprecated` from this test altogether.

mppf reviewed (commented):
I looked over the tests. I also ran testing. 

Next steps for me:
 * I need to look at DistributedBag.chpl itself again
 * I need to run multi-locale testing
 
Next steps for you:
 * consider the test feedback
 * address the failures described below

Looks like some updates will be needed:

`[Error matching .bad file for library/packages/Collection/insertArrayInBag1]`

* you need to update insertArrayInBag1.bad to include the deprecation warning

`[Error matching .bad file for library/packages/Collection/insertArrayInBag2]`

* same here

```
[Error matching program output for library/packages/DistributedBag/add]
[Error matching program output for library/packages/DistributedBag/addBulk]
```

* I'm seeing different output order, where SUCCESS is printed before the warning.

```
[Error matching program output for library/packages/DistributedBag/clear]
[Error matching program output for library/packages/DistributedBag/writeThis]
```

* in these ones, the output depends on the configuration of the system. Can you make the coforall use a fixed number of tasks? I would expect the DistributedBag should work if we oversubscribe? For the number of locales, you can put the number of locales you want it to be tested with in `clear.numLocales` and then add a `clear.skipif` to skip the test for single-locale testing configurations (it could just have `CHPL_COMM=none`). Another possibility is to not print this out but rather check it. For the `writeThis` test, you could write to a `writer` channel from `openMemFile` and then turn that into a string and then do validation on it.


```
[Error matching program output for release/examples/spec/Generics/fully-defaulted]
```

I suspect this is an issue with the particular version of `main` that this branch started from. If you are able to `git merge` or `git rebase` to bring in the changes from `main`, it should address it. If that seems like a challenge, I can do this when I try testing next.

mppf commented on modules/packages/DistributedBag.chpl:
```diff
@@ -326,958 +351,848 @@ module DistributedBag {
     /*
-      Insert an element to this node's bag. The ordering is not guaranteed to be
+      Insert an element in segment ``taskId``. The ordering is guaranteed to be
       preserved.
```
Let's add something here that says that the insertion might fail if the distributed bag is too full, and in that case the function will return `false`

Guillaume-Helbecque commented:
Thank you for this review! I addressed your feedbacks, and fixed test failures. Please find below some comments:
- For `library/packages/DistributedBag/clear`, I adapted the test to make it independant from the system. Now, it simply checks that the bag contains the good number of elements before `clear`, and 0 after. The failure was caused by the printing, which is not usefull in fact, since we already have the `writeThis` test.
- For `library/packages/DistributedBag/writeThis`, I disabled single-locale testing, and fixed the number of tasks in `coforall` loops by setting `CHPL_RT_NUM_THREADS_PER_LOCALE=4` in `EXECENV`, and 3 in `writeThis.numlocales`.
- For the sporadic failure of `library/packages/DistributedBag/nqueens`, it was related to the termination detection. In few words, the algorithm ends when `bag.getSize() == 0`, but we cannot check this condition each iteration because it implies expensive atomic operations. Therefore, I added some lightweight logic around this (i.e. `eachLocaleState`, `eachTaskState`), and it seems that I missed something in the first try.

I also provided some comments directly in your review feedbacks.

mppf reviewed (approved):
I've run comm=none and comm=gasnet oversubscribed testing & they both came back clean. I have no further concerns about this PR but it will have to wait until after 1.33 is released for us to merge it.

Guillaume-Helbecque commented:
OK to merge the PR after the 1.33 release. I just addressed your last feedbacks. 
I'm really excited to get this effort in the package modules library. Thank you again for your time and the valuable reviews!

Guillaume-Helbecque commented:
@mppf: Thanks a lot for merging and for your suggestion on the future work.

bradcray commented:
Nice!  Thanks very much @Guillaume-Helbecque for sticking with this and getting it merged, and to @mppf for shepherding it in!

mppf commented:
@Guillaume-Helbecque 

We're seeing timeouts in nightly testing for `test/library/packages/DistributedBag/nqueens.chpl `.

I've been able to reproduce in two ways.

This configuration is like a nightly testing configuration that failed:

<details>

https://chapel-lang.org/docs/developer/bestPractices/Sanitizers.html#best-practices-sanitizers

```
source util/setchplenv.bash
export CHPL_COMM=gasnet
export CHPL_MEM=cstdlib
export CHPL_TASKS=fifo
export CHPL_LLVM=none   # or set CHPL_TARGET_COMPILER to gnu, clang, etc.
export CHPL_SANITIZE_EXE=address
export ASAN_OPTIONS="use_sigaltstack=0,detect_leaks=0"
make
```

```
./util/start_test  test/library/packages/DistributedBag/nqueens.chpl 
```

We are seeing a timeout in this configuration. 

</details>

This configuration is much simpler and also fails:

<details>

There is an even simpler configuration that shows the problem:

```
source util/quickstart/setchplenv.bash
export CHPL_COMM=gasnet
make
```

```
./util/start_test  test/library/packages/DistributedBag/nqueens.chpl 
```

</details>

There are also warnings printed out: `warning: Setting number of threads in CHPL_TASKS=fifo can lead to deadlock`. Note that your test directory includes an `EXECENV` that sets `CHPL_RT_NUM_THREADS_PER_LOCALE=4` that you added to make `writeThis` have the same output on different systems.

I suspect that the most sensible answer is to move `EXECENV` to `writeThis.execenv`.

Would you be able to investigate and prepare a PR that resolves the issue somehow? Thanks.

Guillaume-Helbecque commented:
@mppf: Yes, I'm going to fix this.

I was able to reproduce with the first script. It seems that `EXECENV` was the problem, as you mentioned. I still have the following warnings, but I guess it is related to the configuration and my system.
```
> warning: Disabling --cache-remote due to incompatibility with AddressSanitizer (quiet with CHPL_RT_CACHE_QUIET=true)
> nqueens.chpl:165: error: halt reached - Too many callers to barrier()
```

I'm going to perform some more verifications, and create a PR for that.

mppf commented:
@Guillaume-Helbecque - we are also seeing a memory leak reported for `nqueens.chpl`. The leak is for this line:

```
  const PrivateSpace: domain(1) dmapped privateDist();
```

This is from a known issue with `privateDist`. The recommended fix is to simply delete this line. That will cause it to use the `PrivateSpace` already declared in the `PrivateDist` module. Could you include this change in the PR you are working on? Thanks.

Guillaume-Helbecque commented:
PR #24159 fixes both failures. Thanks

jhh67 commented:
There are still nightly testing time-outs under the following configurations:

- chapcs-correctness-test-gasnet-everything
- chapcs-correctness-test-valgrindexe
- correctness-test-cygwin64

Guillaume-Helbecque commented:
@jhh67: Thanks for reporting, I'm going to investigate this in a separate PR. In the meanwhile, can you give me some more information about the failures (files concerned, any warnings,etc)? How can I reproduce these failures?

mppf commented:
@jhh67 - chapcs-correctness-test-gasnet-everything looks like it is passing now to me? Are you observing a sporadic timeout?

For the valgrind configuration, I think running the nqueens under valgrind just takes too long. I think we can just skip. We will have some valgrind testing of the library in other tests.

I think we can also skip it under cygwin. It's not a high priority platform & I don't think it's worth investigating the issue. It could be something about the infrastructure.

@Guillaume-Helbecque - I'll make a PR to add these two skips, but if the gasnet configuration is failing sporadically, that'll be something we'll need you to look at. Thanks.

mppf commented:
@Guillaume-Helbecque - FYI, we are still seeing some sporadic timeouts with the new nqueens test. I think there must be a low probability live lock / dead lock. We'll try to gather information about it but at the moment I don't have reproducing instructions.

I adjusted testing to skip the nqueens test in the valgrind configuration, but that's one configuration that has reliably timed out. But I don't know if it's the same problem. We have some notes about using valgrind in https://chapel-lang.org/docs/developer/bestPractices/Valgrind.html . It's also possible that `CHPL_TASKS=fifo` makes the timeout more likely but I'm not sure of that, either.

bradcray commented:
Valgrind is also just slow, so that timeout didn't necessarily seem surprising to me.

@Guillaume-Helbecque, I'm curious whether you think it's more likely for an issue to be in nqueens or in the DistributedBag itself.  Have you seen hangs in other codes like ChOp that use your DistributedBag?

Guillaume-Helbecque commented:
> Valgrind is also just slow

In this N-queens test, we solve the instances from 1 to `maxN` (12 by default). If some system configurations are slow, decrease `maxN` might help.

> I'm curious whether you think it's more likely for an issue to be in nqueens or in the DistributedBag itself. Have you seen hangs in other codes like ChOp that use your DistributedBag?

As far as I remember, I never had this kind of timeout in my use cases (UTS, N-Queens, Flowshop scheduling, 0/1 Knapsack). However, I always considered `CHPL_TASK=qthread` without over-subscription. What I can do in the next days is to execute my own code on different configurations and see if any problem comes.

bradcray commented:
> If some system configurations are slow, decrease maxN might help.

The case that we saw went from 16s on the previous nights to 1800s (at which point it was killed) on the night it timed out, which makes it feel like more than just a too-big problem size.  This is using qthreads, but it is using oversubscription...  Running 3 locales on an M1 Mac.  The environment was:

```
> CHPL_HOST_PLATFORM: darwin *
> CHPL_HOST_COMPILER: clang
>  CHPL_HOST_CC: clang
>  CHPL_HOST_CXX: clang++
> CHPL_HOST_ARCH: arm64
> CHPL_TARGET_PLATFORM: darwin
> CHPL_TARGET_COMPILER: llvm
>  CHPL_TARGET_CC: /opt/homebrew/Cellar/llvm@15/15.0.7/bin/clang
>  CHPL_TARGET_CXX: /opt/homebrew/Cellar/llvm@15/15.0.7/bin/clang++
>  CHPL_TARGET_LD: /opt/homebrew/Cellar/llvm@15/15.0.7/bin/clang++
> CHPL_TARGET_ARCH: arm64
> CHPL_TARGET_CPU: none *
> CHPL_LOCALE_MODEL: flat
> CHPL_COMM: gasnet *
>  CHPL_COMM_SUBSTRATE: udp
>  CHPL_GASNET_SEGMENT: everything
>  CHPL_GASNET_VERSION: 1
> CHPL_TASKS: qthreads
> CHPL_LAUNCHER: amudprun
> CHPL_TIMERS: generic
> CHPL_UNWIND: none
> CHPL_HOST_MEM: cstdlib
> CHPL_MEM: jemalloc
> CHPL_ATOMICS: cstdlib
>  CHPL_NETWORK_ATOMICS: none
> CHPL_GMP: bundled
> CHPL_HWLOC: bundled
> CHPL_RE2: bundled
> CHPL_LLVM: system
>  CHPL_LLVM_SUPPORT: system
>  CHPL_LLVM_CONFIG: /opt/homebrew/opt/llvm@15/bin/llvm-config
>  CHPL_LLVM_VERSION: 15
> CHPL_AUX_FILESYS: none
> CHPL_LIB_PIC: none
> CHPL_SANITIZE: none
> CHPL_SANITIZE_EXE: none
```

Of all those factors, the "M1" (arm-based) aspect gives me the most pause since there could be a synchronization bug that is potentially specific to that running on that processor.  If that's the case, the problem is more likely ours or Qthreads than this code's—unless it's a MCM bug in the package or test that is not problematic on x86 (which would explain why we haven't seen other similar timeouts in M1 testing, at least that I can recall).

My current opinion is that we should maybe not worry too much about this and gather more data to see how prevalent it is, and whether it shows up in other configurations over time.  I.e., if nothing jumps out at you as being obviously problematic (and given that you haven't seen similar issues in your own ChOp runs over time), I don't know that this is worth more than some casual musing at present.

If it does occur frequently, then we'll need to figure out a way to quiet the failures.  And if it seems to be specific to the M1 configurations, that might suggest that we should do more investigation on our own side.

mppf commented:
We're still seeing the issue periodically. We've also observed the issue with x86-64 and oversubscribed GASNet with segment=fast. However, I haven't been able to reproduce it myself yet. On my system, the test takes 5-20 seconds in that configuration.

Regarding the valgrind configuration, on my system the test took about 410 seconds, but it did complete successfully.

<details>

I used this configuration

```
source util/quickstart/setchplenv.bash
make
./util/start_test  test/library/packages/DistributedBag/nqueens.chpl  -valgrindexe
```

</details>

vasslitvinov commented:
Hi @Guillaume-Helbecque , I was able to look at stack traces in an nqueens run that was still going after five minutes since it started, vs. the expected couple of seconds. It has 3 locales, 26 threads per locale. A few of the threads have stack traces without compiler-generated code in it, such as qthreads sitting in pthread_cond_timedwait or the polling thread.

The other threads are in one place or another in `bag.getSize()` or `allIdle()`, except one thread is in `bag.remove()`, in `nqueens_search()` that to myself I summarize like this:

```chpl
coforall loc in Locales do on loc {
    coforall taskId in 0..#here.maxTaskPar {
        while true {
            var (hasWork, parent) = bag.remove(taskId);
            if hasWork {
                 ......
            }
            else {
                if allIdle(.....) && bag.getSize() == 0 {
                    break; // end of the global search
                }
                continue;
            }
            ...... potentially add to the bag .......
        }
    }
}
```

The execution of `remove()` is at the point of launching these on-tasks in the `REMOVE_GLOBAL_STEAL` case:
```chpl
for victimLocaleId in victim(numLocales, here.id, policy, 1) {
  on Locales[victimLocaleId] {
    var targetBag = chpl_getPrivatizedCopy(parentHandle.type, parentPid).bag;
    .......
```

My working hypothesis is that `remove()` keeps returning `false` while `getSize()` keeps returning non-0. This, in turn, may be because some elements ended up non-steal-able, however still there.

I am posting it here as an update and to see if you have any quick reactions / aha moments / ... I hope to have more info soon.
